hivemind/utils/grpc.py

```diff
@@ -160,6 +161,45 @@ def top(self) -> ValueError:
         raise ValueError(f"Please use {self.__class__.__name__}.get_stub to get or create stubs")
 
 
+def quantile_encode_torch_qq(weight, n_bits: int):
+    n_bins = 2 ** n_bits
+    borders = torch.as_tensor(quantile_qq_approximation(weight.numpy(), n_bins + 1)[:-1])
+    quant_weight = torch.relu_(torch.bucketize(weight, borders) - 1)
+    _sums = torch.zeros(n_bins).scatter_add_(0, quant_weight.flatten(), weight.flatten())
+    lookup = _sums.mul_(n_bins / weight.numel())
```

Comment:
note: it may be better to divide not by weight.numel(), but by the actual number of values in each bin